File: trabalho1/main.cpp
#include <iostream>
#include <pthread.h>
#include "../../BlackGPIO/BlackGPIO.h"
#include "../../ADC/Adc.h"
#include "../../PWM/PWM.h"
#include <unistd.h>

#define PERIODO 500000

using namespace std;
using namespace BlackLib;

//Espera botão ser pressionado por pelo menos 200ms e ser solto
bool isPressed(BlackGPIO &botao){
    if (botao.isHigh()) {
        cout << "BOTÃO PRESSIONADO" << endl;
        usleep(200000);
        if(botao.isHigh()) { 
            //Aguar botão ser "despressionado para emitir sinal de verdadeiro"     
            while(botao.isHigh()){}
            cout << "BOTÃO SOLTO" << endl;
            return true;
        }
        else{
            cout << "BOTÃO SOLTO" << endl;            
        }
    }
    return false;
}

int main()
{
    // Inicialização dos pinos GPIO, ADC e PWM
    BlackGPIO botao(GPIO_60, input);          
    BlackGPIO led(GPIO_30, output);           
    ADC adc(AIN0);
    PWM pwm(P9_22);
    pwm.setState(statePwm::run);
    float valorADC = 0;
    
    //Bloqueia programa enquanto botão não for pressionado
    while (!isPressed(botao)){}
    cout << "ACENDENDO LED" << endl;
    led.setValue(high);	

    while (!isPressed(botao)){
	    valorADC = adc.getPercentValue();
        cout << "ADC (%): " << valorADC << "; ";
        pwm.setPeriod(PERIODO); //periodo de 500000 ns
        pwm.setDutyCycle(PERIODO*valorADC/100.0);
        cout << "DC PWM (ns): " << pwm.getDutyCycle() << endl;
    }

    cout << "DESLIGANDO LED" << endl;
    led.setValue(low);
    cout << "DESLIGANDO PWM" << endl;
    pwm.setDutyCycle(0);
    pwm.setState(statePwm::stop);
    cout << "FIM" << endl;
    return 0;
}

File: trabalho1/main.py
#coding:utf-8
import Adafruit_BBIO.GPIO as GPIO
import Adafruit_BBIO.PWM as PWM
import Adafruit_BBIO.ADC as ADC
import time

#Definindo pinos
PIN_LED = "P9_11"
PIN_BUTTON = "P9_12"
PIN_PWM = "P9_22"
PIN_ADC = "P9_39"

#Configurando os pinos
ADC.setup()                         #Configurando ADC
GPIO.setup(PIN_LED, GPIO.OUT)       #Configurando o pino do LED com saída
GPIO.setup(PIN_BUTTON, GPIO.IN)     #Configurando o pino do botão como entrada
PWM.start(PIN_PWM, 0)               #Configurando PWM com duty cycle = 0

#Funcão que verifica se o botao foi pressiondo por pelo menos 0.2 segundos
def isPressed(pin):
    if GPIO.input(pin):
        print("BOTÃO PRESSIONADO")
        time.sleep(0.2)
        if GPIO.input(pin):
            while GPIO.input(pin):
                pass
            print("BOTÃO SOLTO")
            return True
        else:
            print("BOTÃO SOLTO")
    return False


#Se botão for pressionado, Led acende e PWM = ADC*100
while not isPressed(pin=PIN_BUTTON):
    pass
print("ACENDENDO LED")
GPIO.output(PIN_LED, GPIO.HIGH)
           
while not isPressed(pin=PIN_BUTTON):                
    value = ADC.read(PIN_ADC)
    PWM.set_duty_cycle(PIN_PWM, value*100.0)
    print("ADC: " + str(value) + "; PWM (%): " + str(value*100.0)) #Uma vez que não há uma função para obter o valor do PWM
        
#Define 0 como saída do pino e desativa GPIO
print("DESLIGANDO LED")
GPIO.output(PIN_LED, GPIO.LOW)
GPIO.cleanup()
#Pára PWM e o desativa 
#PWM.set_duty_cycle(PIN_PWM, 0) 
print("DESLIGANDO PWM")
PWM.stop(PIN_PWM)
PWM.cleanup()
print("FIM DO PROGRAMA")
